Add support for Open Policy Agent

vagaerg opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

This PR adds support for using OPA as an access control in Trino. It is forked off from the great work here: https://github.com/stackabletech/trino-opa-authorizer (CC @sbernauer @nightkr @maltesander @lfrancke @fhennig )

Also note that additional discussion can be had on slack if needed. We have a openpolicyagent-dev channel to facilitate this. Ping @mosabua or @colebow to be added.


<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

We're opening this to get feedback and discuss improvements. This is not fully complete, a few items remain:

- [ ] Better documentation (for now, refer to the README)
- [ ] Adding sample rego policies in the documentation
- [x] Using the Airlift HTTP client instead of the `java.net` one
- [x] Using the JsonMapper from Airlift instead of building our own
- ~~[ ] Moving tests to TestNG so they match the main Trino codebase~~
- [x] Containerize OPA or otherwise make the binary available to the test suite

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(X) Release notes are required, with the following suggested text:

```markdown
# Add Open Policy Agent plugin
* Implement OPA plugin in Trino mainline ( #9787 )
```

mosabua commented:
Excellent. We probably need to confirm that all committers on that other repository also signed the CLA and are okay with this progress here. Could you mention them here @vagaerg as a first step and maybe ideally they can each confirm as a separate comment.

vagaerg commented:
> Excellent. We probably need to confirm that all committers on that other repository also signed the CLA and are okay with this progress here. Could you mention them here @vagaerg as a first step and maybe ideally they can each confirm as a separate comment.

They were mentioned on the first PR comment -  @sbernauer @nightkr @maltesander @lfrancke @fhennig . I've been in touch over email too, hopefully they'll see this soon!
Thanks

soenkeliebau commented:
Thank you very much for this @vagaerg !!
I look forward to checking out what you added!

Regarding the CLA, Ianal - but I believe that no CLA is required, since that code is released by Stackable under the Apache 2.0 license.

Praveen2112 commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizerFactory.java:
```diff
@@ -0,0 +1,45 @@
+
+        Bootstrap app = new Bootstrap(new OpaAuthorizerModule());
+
+        Injector injector = app.doNotInitializeLogging().setRequiredConfigurationProperties(config).initialize();
```
nit: Can we move `.doNot..`, `.setRequired` and `.initialize` to new lines ?

Praveen2112 commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizerModule.java:
```diff
@@ -0,0 +1,58 @@
+        configBinder(binder).bindConfig(OpaConfig.class);
+        install(conditionalModule(
+                OpaConfig.class,
+                config -> config.getOpaBatchUri() != null,
```
nit: How about using `Optional` here.

Praveen2112 commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaBatchAuthorizer.java:
```diff
@@ -0,0 +1,128 @@
+{
+    private final URI opaBatchedPolicyUri;
+
+    public static class OpaBatchQueryResult
+    {
+        @JsonProperty("decision_id")
+        public String decisionId;
+        public List<Integer> result;
+    }
```
Can we move it to a record ?

Praveen2112 commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaQueryInput.java:
```diff
@@ -0,0 +1,28 @@
+
+import io.trino.spi.security.SystemSecurityContext;
+
+public class OpaQueryInput
+{
+    public final SystemSecurityContext context;
+    public final OpaQueryInputAction action;
+
+    public OpaQueryInput(SystemSecurityContext context, OpaQueryInputAction action)
+    {
+        this.context = context;
+        this.action = action;
+    }
```
Can we move this to `record`

Praveen2112 commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizerFactory.java:
```diff
@@ -0,0 +1,45 @@
+        Bootstrap app = new Bootstrap(new OpaAuthorizerModule());
+
+        Injector injector = app.doNotInitializeLogging().setRequiredConfigurationProperties(config).initialize();
+        return injector.getInstance(Key.get(SystemAccessControl.class, ForOpa.class));
```
Do we need this annotation today ?

vagaerg replied:
Yes, this is used to inject a specific instance of the OPA Access Control (either the `batch` one or the main one)

Praveen2112 reviewed (commented):
Thanks for working on it. Just skimmed it.

sbernauer commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizer.java:
```diff
@@ -0,0 +1,827 @@
+    public OpaAuthorizer(OpaConfig config, HttpClient httpClient)
+    {
+        this.opaPolicyUri = config.getOpaUri();
+        this.json = new ObjectMapper();
+        // do not include null values
+        this.json.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+        // deal with Optional<T> values
+        this.json.registerModule(new Jdk8Module());
```
```suggestion
        this.json = new ObjectMapper()
                // OPA server can send other fields, such as `decision_id` when enabling decision logs
                // We could add all the fields we *currently* know, but it's more future-proof to ignore any unknown fields.
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                // Do not include null values
                .setSerializationInclusion(JsonInclude.Include.NON_NULL)
                // Deal with Optional<T> values
                .registerModule(new Jdk8Module());
```

vagaerg replied:
The latest commit no longer uses the ObjectMapper and instead relies on the airlift JSON module (which is based on jackson too)

sbernauer commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAuthorizerSystemTest.java:
```diff
@@ -0,0 +1,321 @@
+            throws IOException, InterruptedException
+    {
+        InetSocketAddress opaSocket = findAvailableTcpPort();
+        opaServer = new ProcessBuilder(System.getenv().getOrDefault("OPA_BINARY", "opa"), "run", "--server", "--addr",
+                opaSocket.getHostString() + ":" + opaSocket.getPort()).inheritIO().start();
```
```suggestion
        String opa_endpoint = opaSocket.getHostString() + ":" + opaSocket.getPort();
        System.out.println("OPA has endpoint " + opa_endpoint);
        opaServer = new ProcessBuilder(System.getenv().getOrDefault("OPA_BINARY", "opa"),
                "run",
                "--server",
                "--addr", opa_endpoint,
                "--set", "decision_logs.console=true"
         ).inheritIO().start();
```

sbernauer reviewed (commented):
Many thanks for moving this forward!
I suggest back-porting the fix from https://github.com/stackabletech/trino-opa-authorizer/pull/27

ksobolew commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizer.java:
```diff
@@ -0,0 +1,827 @@
+    }
+
+    @Override
+    public void checkCanSetViewAuthorization(SystemSecurityContext context, CatalogSchemaTableName view, TrinoPrincipal principal)
```
Setting view authorization is a thorny subject. It's surprisingly tricky to do in a way which avoids potential data leaks. The recently merged trinodb/trino#16691 is a result of almost two years of figuring out what to do with this :) Here we have something similar to what we did for file-based access control, so probably no action is required, but that's something to definitely keep in mind.

ksobolew replied:
Well, it depends, really, on how much flexibility we want to give. In the file-based access control in Trino we decided to add a new, specific privilege for changing authorization. In others it may require full admin access. We also discussed rules like "setting authorization to USER would require a privilege to impersonate that user, while setting it to ROLE would require a role grant with grant option" and enforcing them in the Trino engine; we eventually decided against the latter and to let each implementation choose its own way, instead. So I guess the way it is now gives maximum flexibility, but minimum guidance in avoiding pitfalls :) I'm just saying that this is a place where you have to be very careful.

vagaerg replied:
Got it. Should these nuances perhaps be documented on the javadoc of the SPI? Just to ensure future implementers are aware

ksobolew replied:
I guess it should. cc: @kokosing

ksobolew commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizer.java:
```diff
@@ -0,0 +1,827 @@
+    @Override
+    public void checkCanCreateViewWithSelectFromColumns(SystemSecurityContext context, CatalogSchemaTableName table, Set<String> columns)
+    {
+        // This refers to a Table, so the resource should be a table and not a view
```
A view can query another view. Why is there a distinction between views and tables anyway? In the SQL model they are meant to be basically indistinguishable.

vagaerg replied:
Agreed - and they pretty much are within the code, they both use the `Table` object. I will change this so that the resource is always `table`

ksobolew commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaBatchAuthorizer.java:
```diff
@@ -0,0 +1,128 @@
+        this.opaBatchedPolicyUri = config.getOpaBatchUri();
+    }
+
+    public OpaBatchAuthorizer(OpaConfig config, HttpClient httpClient)
```
`@VisibleForTesting`?

ksobolew commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/FilteringTestHelpers.java:
```diff
@@ -0,0 +1,48 @@
+
+    public static Stream<Arguments> emptyInputTestCases()
+    {
+        Stream<BiFunction<OpaAuthorizer, SystemSecurityContext, Collection>> callables = Stream.of((authorizer, context) -> authorizer.filterViewQueryOwnedBy(context, Set.<Identity>of()), (authorizer, context) -> authorizer.filterCatalogs(context, Set.of()), (authorizer, context) -> authorizer.filterSchemas(context, "some-catalog", Set.of()), (authorizer, context) -> authorizer.filterTables(conte ...
```
Wrap each argument in a separate line

ksobolew commented:
One thing I forgot to ask: do you have any experience with using it in practice? IOW, does it work for you using the model implemented in this PR? Like I said in one of the comments, I'm not very familiar with OPA, so I haven't really reviewed the semantics.

soenkeliebau commented:
> One thing I forgot to ask: do you have any experience with using it in practice? IOW, does it work for you using the model implemented in this PR? Like I said in one of the comments, I'm not very familiar with OPA, so I haven't really reviewed the semantics.

Full disclosure, we have also not extensively used this on production systems, but the general idea behind the model / abstraction layer that the authorizer uses was to map the numerous methods from the Trino authorizer interface to a single rule in Opa/Rego, to avoid having to implement all of these functions 1:1 in the Rego rules.

So basically the idea was to sorta kinda mutate as shown in the examples below (excuse the crude pseudocode):

```
checkCanShowTables(context, schema) -> allow(context, schema, Actions.ShowTables)
checkCanCreateView(context, view) -> allow(context, view, Actions.CreateView)
```

This would allow to implement rules in Rego following a pattern that is similar to what is shown in the rego playground: https://play.openpolicyagent.org/p/WTwDYYLr8t

This allows to keep the rule itself fairly stable, as the actual grants are looked up in the "DATA" that is loaded.

electrum commented:
Note that we're slowly migrating Trino to use JUnit 5, so that is preferred for new modules. See `lib/trino-filesystem` for an example module. We normally use AssertJ in new code, but the JUnit Jupiter assertions are fine.

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizer.java:
```diff
@@ -0,0 +1,827 @@
+        this(config, HttpClient.newHttpClient());
+    }
+
+    public OpaAuthorizer(OpaConfig config, HttpClient httpClient)
```
I want to use the built-in HTTP client that other modules (like HttpEventListener or SalesforceAuthenticator) use, which may make this extra constructor redundant.

I will address the other comments first and then come back to this one

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizer.java:
```diff
@@ -0,0 +1,827 @@
+    @Override
+    public Collection<Identity> filterViewQueryOwnedBy(SystemSecurityContext context, Collection<Identity> queryOwners)
+    {
+        return queryOwners.parallelStream().filter(queryOwner -> queryOpaWithSimpleResource(context, "FilterViewQueryOwnedBy", new OpaQueryInputResource.Builder().user(new OpaQueryInputResource.User(queryOwner)).build())).collect(Collectors.toSet());
```
Interesting, we took this idea from the original implementation by @sbernauer @nightkr @maltesander @lfrancke @fhennig , as well as the Ranger implementation by @dprophet on #13297

That's not to say this is the right approach of course :)
The main complexity from OPA is that the answer to an authorization query may not be straightforward - the concept of "has any access" is quite vague in OPA because policies are code (rego) and not a static policy document like what you'd see on Ranger or the File Based Access Control.

Whoever writes the rego policy may choose to create a rule to allow access to `FilterTables` based on whether the user can `SelectFromColumns`, but we cannot (IMHO) make that decision for them.

Perhaps `Filter*` is not the right operation name for these cases, but I believe we still need to call the operation something unique (e.g. `HasAnyAccess`) so as to allow policy developers to decide what "having any access" entails.

Unless policies are extremely complex, it should be straightforward to map `HasAnyAccess` or `Filter*` to specific operations within OPA/rego. @sbernauer has suggested that we should perhaps create a set of small rego helpers to make writing policies easier for most users (e.g. by mapping `Filter*` operations to something else) while keeping the specificity in what the plugin sends so users can choose to be very specific if they so desire.

ksobolew replied:
Like I said, I don't know what makes sense for OPA. But whatever it is, it would be best if it helped the implementor in getting the semantics right with minimal effort. Perhaps it can be done generically without writing any additional code in OPA, perhaps not. I'm not going to tell what it should be, though.

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaBatchAuthorizer.java:
```diff
@@ -0,0 +1,128 @@
+    public OpaBatchAuthorizer(OpaConfig config, HttpClient httpClient)
+    {
+        super(config, httpClient);
+        this.opaBatchedPolicyUri = config.getOpaBatchUri();
```
If we use an optional as suggested by @Praveen2112 we can just use `orElseThrow` there right?

ksobolew replied:
Yes, this code is only executed when the `Optional` is not empty, so `orElseThrow` makes sense.

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaQueryInput.java:
```diff
@@ -0,0 +1,28 @@
+
+public class OpaQueryInput
+{
+    public final SystemSecurityContext context;
```
Completely agreed, we've been moving more and more of them but I've missed a few. Will address shortly

ksobolew commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizer.java:
```diff
@@ -0,0 +1,827 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class OpaAuthorizer
```
Somehow, I can't resolve threads in this PR. Feel free to resolve the ones you think are completed.

ksobolew commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaConfig.java:
```diff
@@ -0,0 +1,55 @@
+        return opaUri;
+    }
+
+    public Optional<URI> getOpaBatchUri()
+    {
+        return Optional.ofNullable(opaBatchUri);
+    }
```
It will be better if the field itself is `Optional`.

ksobolew reviewed (commented):
Some more comments. But good work overall!

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaHttpClient.java:
```diff
@@ -0,0 +1,124 @@
+    {
+        this.httpClient = httpClient;
+        this.serializer = serializer;
+        this.executor = new BoundedExecutor(
+                newCachedThreadPool(daemonThreadsNamed("opa-access-control-http-%s")),
+                Runtime.getRuntime().availableProcessors());
```
Just to confirm - We don't need to make it its own `Module` correct? I can use the same binder in OpaAccessControlFactory

vagaerg replied:
Excuse my lack of familiarity here, does the following sound reasonable?

- Bind an `ExecutorService` annotated with `ForOpa` to be used _only_ within this module, allowing us to have a shutdown hook and stop the executor service
- Bind an `Executor` annotated with `ForOpa` to be used by `OpaHttpClient` that retrieves the `ExecutorService` we exposed above and builds  a `BoundedExecutor`
    - (AFAIK there's no `ExecutorService` equivalent of BoundedExecutor)

Also - I've seen some other modules use an optional binder to allow overriding the building of executors, should we allow the same here?

ksobolew replied:
Correct, no need to extract a separate `Module`.

> Bind an ExecutorService annotated with ForOpa to be used only within this module, allowing us to have a shutdown hook and stop the executor service

If by "module" you mean this plugin, then yes, plugins are isolated from each other (by way of using a separate Guice context / `Injector`). Isolation between Guice `Modules` is a different matter (in short: don't rely on it). The binding annotation is most useful in larger Guice contexts, but it makes sense to use it here as well.

> Bind an Executor annotated with ForOpa to be used by OpaHttpClient that retrieves the ExecutorService we exposed above and builds a BoundedExecutor

Right, these are implementations from Airlift, so maybe using `ExecutorService` doesn't make sense indeed. Would binding a `BoundedExecutor` as an `@ForOpa Executor` work for you? The simpler it is the better.

ksobolew commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaHttpClient.java:
```diff
@@ -0,0 +1,124 @@
+    {
+        this.httpClient = httpClient;
+        this.serializer = serializer;
+        this.executor = new BoundedExecutor(
+                newCachedThreadPool(daemonThreadsNamed("opa-access-control-http-%s")),
+                Runtime.getRuntime().availableProcessors());
```
Actually, you may be right, I'm not very familiar with the Airlift HTTP client. I just get uneasy whenever I see a thread pool ;) So this is for non-blocking operations? Is it executed asynchronously? Maybe it's better to just keep things simple, at least at first.

vagaerg replied:
To the best of my knowledge, the Airlift HTTP client maintains its own thread pool and uses it to send the request and perform any operations upon receiving a response.

Airlift HTTP passes a `BufferingResponseListener` to Jetty ([here](https://github.com/airlift/airlift/blob/c6aaa3d9a2d775e986581800611cb67260ea2233/http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java#L747-L754)), which implements an `onComplete` method to process a response once received from the HTTP endpoint.

That method propagates this information to the `onComplete` method of a custom future implemented by Airlift HTTP ([here](https://github.com/airlift/airlift/blob/519fb14dd23b663a80367ebcb6d7b9b871aadc24/http-client/src/main/java/io/airlift/http/client/jetty/BufferingResponseListener.java#L82-L92)). This custom future just deals with processing the complete HTTP response, optionally turning it into something else as controlled by the `ResponseHandler` that was originally passed in when the request was sent (the `handle` method called [here](https://github.com/airlift/airlift/blob/c6aaa3d9a2d775e986581800611cb67260ea2233/http-client/src/main/java/io/airlift/http/client/jetty/JettyResponseFuture.java#L116-L139)); in our case, that is the [JSON parsing handler](https://github.com/airlift/airlift/blob/519fb14dd23b663a80367ebcb6d7b9b871aadc24/http-client/src/main/java/io/airlift/http/client/JsonResponseHandler.java#L68-L87) which deserializes the bytes into a POJO using a `JsonCodec`

My understanding is that this should result in all the I/O blocking *and* JSON parsing taking place on the Airlift HTTP threadpool. Our threadpool would only be used to deal with already received and parsed responses, to do things like filtering them based on specific fields or collecting streams of completed futures into a single collection.

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAccessControlFactory.java:
```diff
@@ -0,0 +1,72 @@
+                binder -> {
+                    jsonCodecBinder(binder).bindJsonCodec(OpaQuery.class);
+                    jsonCodecBinder(binder).bindJsonCodec(OpaQueryResult.class);
+                    if (httpClient.isEmpty()) {
+                        httpClientBinder(binder).bindHttpClient("opa-access-control", ForOpa.class);
+                    }
+                    else {
+                        binder.bind(Key.get(HttpClient.class, ForOpa.class)).toInstance(httpClient.orElseThrow());
+                    }
```
Considering this is only used for testing I thought this would be a reasonable compromise. Happy to change it if you think it's worth the extra complexity though

ksobolew replied:
I prefer to keep testing code as much separated as possible. I know this is not always feasible, though. So I won't be pushing for this, but if you think of some simpler way to make the separation, that would be nice :)

vagaerg replied:
Agreed. All I can think of (from quickly glancing at a few other plugins) is to either:
- Build the authorizer using the constructor instead of relying on the factory and injection (e.g. what the Pinot module does), or
- Spin up a real test HTTP server instead of using a mock HTTP client (e.g. what the HTTP event listener module does)

The former would work, but given the number of objects that we're building through injection now (the `OpaHttpClient`, `OpaHighLevelClient`, etc...) I feel it could get messy. The latter would work but it'd be a larger scale refactor just to remove this line of code.

I'll resolve for now, but if anyone has objections or nicer ideas please reopen :)

vagaerg commented:
> One thing I forgot to ask: do you have any experience with using it in practice? IOW, does it work for you using the model implemented in this PR? Like I said in one of the comments, I'm not very familiar with OPA, so I haven't really reviewed the semantics.

We're pretty much on the same boat as @soenkeliebau . We have not yet used it in production, but following the concept of "action, resource, context" seemed reasonable (and in line with systems like IAM)

We've implemented some rego already and have been happy with this model, but our rego is not generic enough for it to be helpful if we added it here. That being said, I'd be happy to work with @soenkeliebau and anyone else to come up with some helpers to simplify policy creation.

IMHO, the power of OPA is that it allows you to take all the policy decisions outside of Trino - your policies can be as simple or as complex as you wish, no need to change the Trino code. That of course requires Trino to send in very specific information to OPA, as opposed to - say - a more generic version where we could conflate all "read" operations together, all "alter" operations together, etc...

I think sending all this information is useful, but we may want to write some rego helpers to turn this specific operation into more generic rules so that users with simpler needs can write policies quickly without needing to think about all the possible operations that Trino may send into OPA

vagaerg commented:
> Let me know how you want to proceed in terms of documentation. If desired I can send a separate second PR that is in sync with this PR in terms of documentation and we can work towards merging them together. Or you can convert the readme into the documentation yourself as well and keep it in this PR.

So turn it into rst format? Happy to do that if so

I'll also open another PR once this is merged to expand on the documentation and add some sample rego code to help users onboard into OPA

ksobolew commented:
> IMHO, the power of OPA is that it allows you to take all the policy decisions outside of Trino - your policies can be as simple or as complex as you wish, no need to change the Trino code. That of course requires Trino to send in very specific information to OPA, as opposed to - say - a more generic version where we could conflate all "read" operations together, all "alter" operations together, etc...

Okay, that makes sense, and it's fine by me, but I'll let someone more competent in this to also take a look :)

mosabua commented:
I discussed docs with @vagaerg ... given that we are currently migrating docs from rst to md we will work towards merging this PR with the docs in the readme file as they are now .. and then do a follow up PR as soon as possible after the merge of the code.

sbernauer commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/schema/TrinoSchema.java:
```diff
@@ -0,0 +1,58 @@
+    public static class Builder
+            extends BaseSchemaBuilder<TrinoSchema, Builder>
+    {
+        protected Builder getInstance()
```
```suggestion
        @Override
        protected Builder getInstance()
```

sbernauer commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/schema/TrinoTable.java:
```diff
@@ -0,0 +1,75 @@
+        public String tableName;
+        public Set<String> columns;
+
+        protected Builder getInstance()
```
```suggestion
        @Override
        protected Builder getInstance()
```

soenkeliebau commented:
I've lost touch with this a bit over the summer I'm afraid. Trying to understand what the status is on this PR .. it looks to me like all conversations have been resolved?

Can someone help me out? :)

vagaerg commented:
As discussed in Slack, we think this PR is ready for review / merge. We do not have any other immediate changes to make at the moment

sbernauer commented:
@ksobolew would you be available for another round of review? That would be fabulous!

sbernauer commented:
Thanks a lot @ksobolew for taking your time! Even as you are not a maintainer, having someone with relevant knowledge giving a review is always helpful :)

vagaerg commented:
Thanks a lot for the review @ksobolew - will address those!

vagaerg commented:
I have rebased on top of master. Note that there were some changes in the SPI (around `filterColumns`, where we can now get multiple tables at once), so there are a few changes in the way we handle parallel requests.

As discussed over Slack, when getting a call for `filterColumns` with multiple tables, we have opted to:
- Send 1 request per table & column in the non-batched authorizer (similar to what we did before)
- Send 1 request per table, containing all columns within it for the batched authorizer

This decision for the batched authorizer allows us to keep any rego code identical (as it still sends 1 request per table, with all columns inside) and is simpler than trying to send all tables & all columns together IMHO - at the expense of some RTT overhead to the server if there are many tables batched in a single `filterColumns` call

dain commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlFilteringUnitTest.java:
```diff
@@ -0,0 +1,333 @@
+    {
+        this.mockClient = new HttpClientUtils.InstrumentedHttpClient(opaServerUri, "POST", JSON_UTF_8.toString(), (request) -> OK_RESPONSE);
+        this.authorizer = (OpaAccessControl) new OpaAccessControlFactory()
+                .create(Map.of("opa.policy.uri", opaServerUri.toString()),
```
Prefer ImmutableMap (and other immutables) to Java builtin ones

dain commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlFilteringUnitTest.java:
```diff
@@ -0,0 +1,333 @@
+        SchemaTableName tableTwo = SchemaTableName.schemaTableName("my-schema", "table-two");
+        SchemaTableName tableThree = SchemaTableName.schemaTableName("my-schema", "table-three");
+        Map<SchemaTableName, Set<String>> requestedColumns = Map.of(
+                tableOne, Set.of("table-one-column-one", "table-one-column-two"),
```
for short sets like this ImmutableSet.of is fine

dain commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlFilteringUnitTest.java:
```diff
@@ -0,0 +1,333 @@
+                requestedColumns);
+
+        assertEquals(mockClient.getRequests().size(), 0);
+        assertTrue(
+                Maps.difference(
+                        result,
+                        requestedColumns).areEqual());
```
one line

dain commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlFilteringUnitTest.java:
```diff
@@ -0,0 +1,333 @@
+import static io.trino.plugin.openpolicyagent.TestHelpers.NO_ACCESS_RESPONSE;
+import static io.trino.plugin.openpolicyagent.TestHelpers.OK_RESPONSE;
+import static io.trino.plugin.openpolicyagent.TestHelpers.systemSecurityContextFromIdentity;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
```
BTW, we are moving the codebase to asserj's assertThat

vagaerg replied:
Should I do that as part of this PR or would you prefer to do that in a follow up one?

dain replied:
You can do it as a followup.  Just wanted to give you a heads up

dain commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlSystemTest.java:
```diff
@@ -0,0 +1,339 @@
+                Thread.sleep(timeoutMs);
+            }
+        }
+        throw new SocketTimeoutException(
+                "Timed out waiting for addr %s to be available (%d attempts made with a %d ms wait)"
+                        .formatted(addr, attempts, timeoutMs));
```
one line

dain commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlSystemTest.java:
```diff
@@ -0,0 +1,339 @@
+        assertEquals(policyRes.statusCode(), 200, "Failed to submit policy: " + policyRes.body());
+    }
+
+    private Session user(String user)
+    {
+        return testSessionBuilder().setIdentity(Identity.ofUser(user)).build();
+    }
+
+    private Set<String> querySetOfStrings(Session session, String query)
+    {
+        return runner.execute(session, query)
+                .getMaterializedRows()
+                .stream()
+                .map((i) -> i.getField(0).toString())
+                .collect(toImmutableSet());
+    }
+
+    private static Stream<Arguments> filterSchemaTests()
+    {
+        Stream<Pair<String, Set<String>>> userAndExpectedCatalogs = Stream.of(
+                Pair.of("bob", ImmutableSet.of("catalogOne")),
+                Pair.of("admin", ImmutableSet.of("catalogOne", "catalogTwo", "system")));
+        return userAndExpectedCatalogs.map((i) -> Arguments.of(Named.of(i.getFirst(), i.getFirst()), i.getSecond()));
+    }
```
move the helper methods to the bottom of the file

dain commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlUnitTest.java:
```diff
@@ -0,0 +1,1518 @@
+    private static Stream<Arguments> noResourceActionTestCases()
+    {
+        Stream<BiConsumer<OpaAccessControl, SystemSecurityContext>> methods =
+                Stream.of(
```
put `Stream.of` on the previous like and then the indenting won't be as large

dain commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlUnitTest.java:
```diff
@@ -0,0 +1,1518 @@
+                    "decision_id": "foo",
+                    "some_debug_info": {"test": ""}
+                }
+                """, 200));
```
I'd put the second arg on a  new line

dain reviewed (changes requested):
This looks pretty good.  I had lots of style comments, but the actual runtime seems good.  Generally I recommend that you review this code base with the style comments in mind.  

DM me on slack when this is ready for another review or if you have any questions.

vagaerg commented on plugin/trino-open-policy-agent/pom.xml:
```diff
@@ -0,0 +1,182 @@
+        <dependency>
+            <groupId>io.airlift</groupId>
+            <artifactId>http-client</artifactId>
+            <exclusions>
+                <exclusion>
+                    <groupId>io.airlift</groupId>
+                    <artifactId>slice</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.openjdk.jol</groupId>
+                    <artifactId>jol-core</artifactId>
+                </exclusion>
+            </exclusions>
```
Seems like they aren't. Either they were required back when this PR was originally opened or this was a mistake on my side (this was my first interaction with Airlift).

Will remove

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAccessControl.java:
```diff
@@ -0,0 +1,1145 @@
+    @Override
+    public void checkCanSetUser(Optional<Principal> principal, String userName)
+    {
+        // This function is called for every query, asking if the user can become themselves, resulting in e.g.
```
Understood, that makes sense thanks

By not implemented do you mean we should leave the override as a no-op, or that we should remove the implementation and rely on the default implementation on the SPI (which denies access)?

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaHttpClient.java:
```diff
@@ -0,0 +1,181 @@
+        this.executor = executor;
+    }
+
+    public <T> FluentFuture<T> submitOpaRequest(OpaQueryInput input, URI uri, JsonCodec<T> deserializer)
```
If requests fail to send, we'll throw an exception. Any non-200 response from OPA is dealt with in `parseOpaResponse`.

Failed requests are not retried, they are treated in the same way as a deny

dain replied:
In my experience HTTP request fail all the time for various, spurious, reasons.  I suggest adding retries in the future for anything that isn't the server actively saying "I understand what you are asking and access is denied".  This will work around problems like online upgrades where the service temporarily gets redirected to a server that returns 404 for example.  This doesn't need to be done in this commit, but I'd put it on the todo list.

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/schema/OpaQueryResult.java:
```diff
@@ -0,0 +1,30 @@
+public record OpaQueryResult(@JsonProperty("decision_id") String decisionId, Boolean result)
+{
+    @Override
+    public @NotNull Boolean result()
```
I wanted to default the field to `false` if for whatever reason it is not provided on the response from OPA (an "undefined" OPA response without this field may happen in some cases, and we should treat these in the same way as an authorization denied one)

If I make this `boolean` then it clashes with the definition of `Boolean result` on the record definition. That one needs to be the boxed type such that it can be constructed in cases where the JSON response just doesn't contain this field.

To my knowledge (please correct me if I'm wrong), Jackson doesn't have a succinct way of defaulting the value with an annotation, so I either need to:
- Keep this as-is, or
- Use a `JsonSetter` so I can override the value

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/schema/TrinoTable.java:
```diff
@@ -0,0 +1,76 @@
+
+    public static TrinoTable fromTrinoTable(CatalogSchemaTableName table)
+    {
+        return TrinoTable.Builder.fromTrinoTable(table).build();
```
I implemented the logic to create a Builder from the internal Trino classes because it simplifies the creation of these records when some of their fields will be populated from a Trino internal object but others will be populated from somewhere else - e.g. `checkCanSelectFromColumns`: the table information comes from `CatalogSchemaTableName` while the columns are coming from another argument as per the SPI. So while that method may look redundant it comes in handy sometimes.

Using it here makes it (IMHO) a bit cleaner than implementing the logic to call the constructor.

If I call the constructor directly here, I am duplicating the logic that is already in the Builder. It would look something like
```java
    public static TrinoTable fromTrinoTable(CatalogSchemaTableName table)
    {
        return new TrinoTable(
                TrinoSchema.builder()
                        .catalogName(table.getCatalogName())
                        .schemaName(table.getSchemaTableName().getSchemaName())
                        .build(),
                table.getSchemaTableName().getTableName(),
                null);
    }
```

Whereas relying on the `Builder` abstracts away the fact that a `TrinoTable` contains a `TrinoSchema`, as it's all automatically handled by the builders.

Happy to change it if that's more in line with the style, just thought I'd try to clarify first

vagaerg commented on plugin/trino-open-policy-agent/src/test/java/io/trino/plugin/openpolicyagent/OpaAccessControlFilteringUnitTest.java:
```diff
@@ -0,0 +1,333 @@
+                        requestedColumns).areEqual());
+    }
+
+    @ParameterizedTest(name = "{index}: {0}")
+    @MethodSource("io.trino.plugin.openpolicyagent.FilteringTestHelpers#emptyInputTestCases")
```
Got it. Should I rewrite all of them or can we leave them as is for now?

Since we attempt to cover all possible corner cases, there is quite a bit of parameterization at the moment indeed

dain replied:
To clarify, we are moving away from usages of annotations like `@ParameterizedTest` as they are hard to follow and debug.  Lots of test abuse these system to do things like call a test with a boolean to test with and without nulls, when it is way clearer to future readers to just see two tests or a test with two invocations (with and without nulls).  Additionally there are tests that have massive providers (hundreds of cases) and then it is impossible to fix just one of the cases when something changes.  In all of these cases it easier to understand and maintain with explicit cases in the code (not interpolated into the code).

That said there are _rare_ places where this kind of style makes sense, and in those places we recommend that you use a for loop over the data (possibly add an enum for the data).

ksobolew replied:
> To clarify, we are moving away from usages of annotations like `@ParameterizedTest` as they are hard to follow and debug.

(For the record, not everyone agrees with this assessment. But @dain has the final word on it.)

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAccessControl.java:
```diff
@@ -0,0 +1,1145 @@
+        opaHighLevelClient.queryAndEnforce(
+                OpaQueryContext.fromSystemSecurityContext(context),
+                "ShowSchemas",
+                (DenyCallable) AccessDeniedException::denyShowSchemas,
```
This `queryAndEnforce` argument is a callable that will be invoked if authorization is denied by OPA.
`queryAndEnforce` has two overloads as to what this argument can be, it can be either:
- A no-arg callable (that's what `DenyCallable` defines and why I casted here)
- A callable that takes a single argument of whatever type the resource passed to `queryAndEnforce` was (e.g. if we queried it with a `CatalogSchemaName` it'd pass that to the deny)

`denyShowSchemas` can be called with no-args or with a single string argument (`extraInfo`). In this instance we want to ensure the no-arg method is called, hence the cast.

I can alternatively change this for

```java
() -> AccessDeniedException.denyShowSchemas()
```

But it felt unnecessarily verbose.
What do you think?

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAccessControl.java:
```diff
@@ -0,0 +1,1145 @@
+    }
+
+    @Override
+    public void checkCanGrantSchemaPrivilege(SystemSecurityContext context, Privilege privilege, CatalogSchemaName schema, TrinoPrincipal grantee, boolean grantOption)
```
What would OPA need to support in this instance? At the moment the idea is that it sends the request as to whether the privilege can be granted, and if the OPA endpoint returns `true` then the operation is allowed.

Is there anything specific to privilege granting that would require additional logic?

vagaerg commented on plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAccessControl.java:
```diff
@@ -0,0 +1,1145 @@
+    }
+
+    @Override
+    public void checkCanRevokeRoles(SystemSecurityContext context, Set<String> roles, Set<TrinoPrincipal> grantees, boolean adminOption, Optional<TrinoPrincipal> grantor)
```
OPA would support answering the yes/no question as to whether someon can revoke a role. Similar question to your point above on privileges, is there anything specific to roles I should be aware of?

If so, any pointers would be much appreciated!

dain commented on plugin/trino-opa/src/main/java/io/trino/plugin/opa/schema/TrinoColumn.java:
```diff
@@ -0,0 +1,35 @@
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Helper record used to denote a specific column within a schema & table pair
+ * This is only used in filterColumns where simpler objects than TrinoTable help
+ * keep the code cleaner
```
I find this helper to be confusing in the schema package.  It is really a private detail of the OpaAccessControl, and should be a private class in there.  That said, I don't think this helper helps... see my comment in OpaAccessControl

dain commented on plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaAccessControl.java:
```diff
@@ -0,0 +1,978 @@
+                table);
+    }
+
+    @Override
+    public Map<SchemaTableName, Set<String>> filterColumns(SystemSecurityContext context, String catalogName, Map<SchemaTableName, Set<String>> tableColumns)
+    {
+        ImmutableSet.Builder<TrinoColumn> allColumnsBuilder = ImmutableSet.builder();
+        Map<SchemaTableName, ImmutableSet.Builder<String>> resultBuilder = new HashMap<>();
+
+        for (Map.Entry<SchemaTableName, Set<String>> oneTableColumns : tableColumns.entrySet()) {
+            SchemaTableName schemaTableName = oneTableColumns.getKey();
+            resultBuilder.put(schemaTableName, ImmutableSet.builder());
+            for (String columnForTable : oneTableColumns.getValue()) {
+                allColumnsBuilder.add(new TrinoColumn(schemaTableName, columnForTable));
+            }
+        }
+
+        Set<TrinoColumn> filteredColumns = opaHighLevelClient.parallelFilterFromOpa(
+                allColumnsBuilder.build(),
+                tableColumn -> buildQueryInputForSimpleResource(
+                        OpaQueryContext.fromSystemSecurityContext(context),
+                        "FilterColumns",
+                        OpaQueryInputResource.builder()
+                                .table(TrinoTable.builder()
+                                        .catalogName(catalogName)
+                                        .schemaName(tableColumn.schemaTableName().getSchemaName())
+                                        .tableName(tableColumn.schemaTableName().getTableName())
+                                        .columns(ImmutableSet.of(tableColumn.columnName()))
+                                        .build())
+                                .build()));
+
+        for (TrinoColumn filteredColumn : filteredColumns) {
+            resultBuilder.get(filteredColumn.schemaTableName()).add(filteredColumn.columnName());
+        }
+
+        return resultBuilder.entrySet().stream().collect(toImmutableMap(Map.Entry::getKey, (mapEntry) -> mapEntry.getValue().build()));
+    }
```
I think this is much simpler and doesn't require the helper TrinoColumn class:
```
    @Override
    public Map<SchemaTableName, Set<String>> filterColumns(SystemSecurityContext context, String catalogName, Map<SchemaTableName, Set<String>> tableColumns)
    {
        ImmutableSet.Builder<TrinoTable> allColumnsBuilder = ImmutableSet.builder();
        for (Map.Entry<SchemaTableName, Set<String>> entry : tableColumns.entrySet()) {
            SchemaTableName schemaTableName = entry.getKey();
            TrinoTable trinoTable = new TrinoTable(catalogName, schemaTableName.getSchemaName(), schemaTableName.getTableName());
            for (String columnName : entry.getValue()) {
                allColumnsBuilder.add(trinoTable.withColumns(ImmutableSet.of(columnName)));
            }
        }

        Set<TrinoTable> filteredColumns = opaHighLevelClient.parallelFilterFromOpa(
                allColumnsBuilder.build(),
                tableColumn -> buildQueryInputForSimpleResource(
                        OpaQueryContext.fromSystemSecurityContext(context),
                        "FilterColumns",
                        OpaQueryInputResource.builder().table(tableColumn).build()));

        ImmutableSetMultimap.Builder<SchemaTableName, String> results = ImmutableSetMultimap.builder();
        for (TrinoTable tableColumn : filteredColumns) {
            results.put(new SchemaTableName(tableColumn.schemaName(), tableColumn.tableName()), getOnlyElement(tableColumn.columns()));
        }
        return Multimaps.asMap(results.build());
    }
```

vagaerg replied:
Agreed this is simpler, I was trying to avoid building the internal `SchemaTableName` object myself but I imagine that is not a big deal.

A specific behaviour of the current implementation (and explicitly covered by my tests) is the case when a table has 0 columns after filtering. I have a test to assert that the keyset of the result is the same as the keyset of the input, tables with no permisisoned columns would just have an empty set in the map entry.

Is there any guidance as to whether this is needed? If Trino is OK with tables with 0 columns being missing from the map returned by `filterColumns` then your approach looks cleaner indeed

vagaerg replied:
I will for now modify the tests assuming that it is OK for the resulting map returned by these functions to be missing entries that have 0 items after filtering.

If that is in fact not OK please let me know and I will change it back

vagaerg commented on plugin/trino-opa/src/main/java/io/trino/plugin/opa/schema/BaseSchemaBuilder.java:
```diff
@@ -0,0 +1,62 @@
+
+    public B properties(Map<String, ?> properties)
+    {
+        // https://openjdk.org/jeps/269
+        // ImmutableMap along with other new collections does not support null
+        // cast nulls to empty optionals
```
I was unsure as to whether anything could possibly pass them in, so I opted for what I thought was safest by dealing with them if for whatever reason nulls are in the map.
I will remove this for now if you believe this is not needed and we can revisit if we encounter issues

dain replied:
The places where we could have null use `Optional<Object>` for the value.

vagaerg commented on plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaAccessControl.java:
```diff
@@ -0,0 +1,978 @@
+    @Override
+    public void checkCanCreateCatalog(SystemSecurityContext context, String catalog)
+    {
+        opaHighLevelClient.queryAndEnforce(
+                OpaQueryContext.fromSystemSecurityContext(context),
+                "CreateCatalog",
+                AccessDeniedException::denyCreateCatalog,
+                OpaQueryInputResource.Builder::catalog,
+                catalog);
```
I think between these two I would rather the first, there isn't much benefit to the extra complexity of passing a callable to perform operations on the resource builder.

My original intention with passing in the "raw" object and a method reference to a method in the OpaQueryInputResource.Buidler was so we could abstract the deny logic and the query building logic in the same way

Say
```java
        opaHighLevelClient.queryAndEnforce(
                OpaQueryContext.fromSystemSecurityContext(context),
                "CreateCatalog",
                catalog,
                AccessDeniedException::denyCreateCatalog,
                OpaQueryInputResource.Builder::catalog);
```

In this instance, `denyCreateCatalog` can get the `catalog` argument directly, and so can `OpaQueryInputResource.Builder::catalog`.

My idea didn't quite work out as you pointed out, since this isn't always possible and there are so many possible combinations that it exploded the number of overloads we needed to implement

I will change to the first style, and will also create small helpers where needed directly in `OpaAccessControl` to try to reduce duplication

vagaerg commented on plugin/trino-opa/src/main/java/io/trino/plugin/opa/schema/TrinoTable.java:
```diff
@@ -0,0 +1,95 @@
+import static java.util.Objects.requireNonNull;
+
+@JsonInclude(NON_NULL)
+public record TrinoTable(
```
Thanks for the input! I have made the change, agreed given the simplicity of this class it is probably not worth it having all that added complexity.

The only case where I have left the resource composition in place is in places like `TrinoFunction` that optionally take a catalog *and* a schema or neither of them. In these cases it is easier to use composition of simpler types so we can reuse the `requireNotNull` that's already implemented in each object

vagaerg commented on plugin/trino-opa/src/main/java/io/trino/plugin/opa/schema/TrinoIdentity.java:
```diff
@@ -0,0 +1,66 @@
+public record TrinoIdentity(
+        @NotNull String user,
+        @NotNull Set<String> groups,
+        @NotNull Set<String> enabledRoles,
+        @NotNull Map<String, OpaSelectedRole> catalogRoles,
+        @NotNull Map<String, String> extraCredentials)
```
That makes sense, I wasn't familiar with the first two. Will remove them

We have played around (just for demos, nothing we have actually deployed) with the idea of using extra credentials to pass in JWTs that grant users extra permissions that are validated via OPA. Do you know what would be a better alternative to do this without using extraCredentials?

vagaerg commented:
Will open a new PR now
